### PR TITLE
Add simple argument-only lifted nn.grad function.

### PR DESCRIPTION
**Pull request by levskaya**
This function only peforms a lifted value-and-grad operation with respect to the arguments of a function, and does not try to calculate gradients with respect to any variables.  This mirrors the behavior of the haiku grad function, and also easily works in the multi-scope setting (external modules passed in) while avoiding the complexities associated with that case for a more general vjp.

### CODE REVIEW

File: flax/linen/transforms.py [hunk 1]
```diff
@@ -1130,6 +1132,95 @@ def __call__(self, x, y):
...
+    If ``has_aux`` is ``False``, returns a ``primals_out, grads`` pair, where
+    ``primals_out`` is ``fn(*primals)``.  ``grads`` are the gradients for the
+    corresponding primals and do not include the gradients for module variables.
+    If ``has_aux`` is ``True``, returns a
+    ``(primals_out, aux), grads`` tuple where ``aux`` is the auxiliary data
+    returned by ``fn``.
+  """
+
+  @functools.wraps(fn)
+  def grad_fn(mdl, *primals):
```
Comment (cgarciae) [inline flax/linen/transforms.py:1198]: why do we create `grad_fn` and immediately use it below?

File: flax/linen/transforms.py [hunk 2]
```diff
@@ -1130,6 +1132,95 @@ def __call__(self, x, y):
...
+def grad(
+  fn: Callable[..., Any],
+  mdl: Module,
+  *primals,
+  has_aux: bool = False,
+  reduce_axes=(),
+  variables: lift.CollectionFilter = True,
+  rngs: lift.PRNGSequenceFilter = True,
+):
+  """A limited, lifted equivalent of ``jax.value_and_grad``.
```
Comment (cgarciae) [inline flax/linen/transforms.py:1144]: Should we call this `value_and_grad` then for consistency with `jax` APIs?

File: flax/linen/transforms.py [hunk 3]
```diff
@@ -1130,6 +1132,95 @@ def __call__(self, x, y):
...
+      reduce_axes=reduce_axes,
+      vjp_variables=False,
+      variables=variables,
+      rngs=rngs,
+      multi_scope=True,
+    )
+
+    if has_aux:
+      out, vjp_fun, aux = vjp_partial()
+      _, *argument_grads = vjp_fun(jax.numpy.ones_like(out))
```
Comment (cgarciae) [inline flax/linen/transforms.py:1214]: Don't we get an additional dimension compared to `grad` here? Maybe we need to `squeeze` the additional dim?

Comment (levskaya) [reply flax/linen/transforms.py:1214]: huh? I don't think so?  why would it add dimensions?  you can double-check the tests

File: flax/linen/transforms.py (hunk 1 above)
Comment (levskaya) [reply flax/linen/transforms.py:1199]: this is how the "direct" vjp/jvp transforms work - they directly call the function.

Comment (levskaya) [reply flax/linen/transforms.py]: oh lol, I see what you mean - it's completely pointless here, yeah I'll fix it.

**Review by cgarciae — APPROVED** [review]
